### PR TITLE
2 Map Trigger Actions - Fire SW at

### PR DESCRIPTION
**Pull request by FS-21**
"Fire SW at location" (Action 505) and "Fire SW at Waypoint" (Action 506) let you fire with a SW from the [SuperWeaponTypes] at a specified coordinate X & Y or at a waypoint.

### CODE REVIEW

File: src/Ext/TAction/Body.cpp [hunk 1]
```diff
@@ -255,6 +261,178 @@ bool TActionExt::BinaryOperation(TActionClass* pThis, HouseClass* pHouse, Object
...
+}
+
+bool TActionExt::RunSuperWeaponAt(TActionClass* pThis, int X, int Y)
+{
+	if (SuperWeaponTypeClass::Array->Count > 0)
+	{
+
+		int swIdx = pThis->Param3;
+		int houseIdx = -1;
+		DynamicVectorClass<int> housesListIdx;
```
Comment (Otamaa) [inline src/Ext/TAction/Body.cpp:299]: Please use `std::vector<int>` instead for temporary vector usage for better resource management.

File: src/Ext/TAction/Body.cpp [hunk 2]
```diff
@@ -255,6 +261,178 @@ bool TActionExt::BinaryOperation(TActionClass* pThis, HouseClass* pHouse, Object
...
+		CellStruct targetLocation;
```
Comment (Otamaa) [inline src/Ext/TAction/Body.cpp:300]: `CellStruct targetLocation =  { (short)X,(short)Y } ;`

File: src/Ext/TAction/Body.cpp [hunk 3]
```diff
@@ -255,6 +261,178 @@ bool TActionExt::BinaryOperation(TActionClass* pThis, HouseClass* pHouse, Object
...
+				houseIdx = pThis->Param4;
+			else
+				return true;
+
+			break;
+		}
+
+		HouseClass* pHouse = HouseClass::Array->GetItem(houseIdx);
+		SuperWeaponTypeClass* pSuperType = SuperWeaponTypeClass::Array->GetItem(swIdx);
+		SuperClass* pSuper = new SuperClass(pSuperType, pHouse);
```
Comment (Otamaa) [inline src/Ext/TAction/Body.cpp:424]: dont use `new` !  use   `GameCreate<SuperClass>(args...)`

**Comment by FS-21** [comment]
For testing the new Trigger actions put these sections in any map. If the sections exist then only add in each section the lines "0123456... ":

```
[Tags]
01234561=2,New trigger 1,01234567

[Triggers]
01234567=Neutral,<none>,Test of a new action,0,1,1,1,0

[Events]
01234567=1,13,0,30

[Actions]
01234567=1,505,0,0,<SW_index>,<House_index>,<Coord_X>,<Coord_Y>,A
```


I'll explain each <...>:

<Coord_X> and <Coord_Y> are cell coordinates, look in FinalAlert 2 for picking the desired location for the test.
if you use the value in -1 in X and/or Y it will pick a random value for X and/or Y.

<House_index> is basically the players locations that will play in the map + Neutral House + Special House.
- If the player doesn't exist then the Trigger won't fire.
- In skirmish games Player "0" is the human player.
- You can set is -1 and the owner of the Super Weapon will be a random player (Neutral houses are ignored here).
- You can set -2 and the owner of the Super Weapon will be Neutral House.
- You can set -3 and the owner of the Super Weapon will a random Human Player. This makes sense if the test is done in multiplayer games.

<SW_index> is the index of the Super Weapon from the rulesmd.ini section called [SuperWeaponTypes]

**EXAMPLES:**

Example 1:
[Actions]
; Fire NukeSpecial (in my case is index "0") as Neutral House at location X=50 & Y=50
01234567=1,505,0,0,0,-2,50,50,A

Example 2:
[Actions]
; Fire NukeSpecial as a random Human Player at location X=random & Y=50
01234567=1,505,0,0,0,-3,-1,50,A

Example 3:
[Actions]
; Fire LightningStormSpecial (in my case is index "2") as a random Player at location X=random & Y=random
01234567=1,505,0,0,2,-1,-1,-1,A

Example 4:
[Actions]
; Fire ParaDropSpecial (in my case is index "5") as a random Player at location X=random & Y=random
01234567=1,505,0,0,2,-1,-1,-1,A

Example 5:
[Actions]
; Fire LightningStormSpecial (in my case is index "2") as a Player "0" (or "4475") at location X=24 & Y=50
01234567=1,505,0,0,5,0,24,50,A

Example 6 (this time with Waypoints):
[Actions]
; Fire LightningStormSpecial (in my case is index "2") as a random Player at Waypoint 7**
01234567=1,506,0,0,2,-1,7,0,A
; ** be sure to set the [Waypoints] with the desired location. For example: 7=100100 for X=100 & Y=100

**Comment by ayylmaoRotE** [comment]
Works well. Tested multiple functions above and can confirm working as described. Would be nice to include a line for FAData.ini too.

**Review by Otamaa — APPROVED** [review]
Look good , we can merge this

File: src/Ext/TAction/Body.cpp [hunk 4]
```diff
@@ -255,6 +261,175 @@ bool TActionExt::BinaryOperation(TActionClass* pThis, HouseClass* pHouse, Object
...
+		{
+			if (X < 0)
+			{
+				targetLocation.X = (short)ScenarioClass::Instance->Random.RandomRanged(0, MapClass::Instance->MapCoordBounds.Right);
+			}
+
+			if (Y < 0)
+			{
+				targetLocation.Y = (short)ScenarioClass::Instance->Random.RandomRanged(0, MapClass::Instance->MapCoordBounds.Bottom);
+			}
```
Comment (Metadorius) [inline src/Ext/TAction/Body.cpp:311]: Unnnecassary braces.

**Review by Metadorius — CHANGES_REQUESTED** [review]
Needs FAData.ini entry and to fix compilation.

File: src/Ext/TAction/Body.cpp [hunk 5]
```diff
@@ -255,6 +261,171 @@ bool TActionExt::BinaryOperation(TActionClass* pThis, HouseClass* pHouse, Object
...
+				houseIdx = pThis->Param4;
+			else
+				return true;
+
+			break;
+		}
+
+		HouseClass* pHouse = HouseClass::Array->GetItem(houseIdx);
+		SuperWeaponTypeClass* pSuperType = SuperWeaponTypeClass::Array->GetItem(swIdx);
+		SuperClass* pSuper = GameCreate<SuperClass>(pSuperType, pHouse);
```
Comment (chaserli) [inline src/Ext/TAction/Body.cpp:417]: Can this be `new`ed directly? wouldn't that break the house's Supers array?

**Pull request merged**